File: business-api/bin/wrobo_biz_api/scopes/data_cache.py
"""Data-cache scope: list / create / get / lookup / upsert / import.

Mirrors business-api/src/cli/commands/data-cache.ts. URL shapes come from
business-api/src/routes/data-caches.ts and the mount path in
business-api/src/app.ts (`/api/v1/data-caches`).

The ``import`` subcommand is wired (the server exposes
``POST /api/v1/data-caches/<slug>/import``) but it does its file reading
client-side: the local CLI parses the file, then calls the service. The
wrapper does the same. CSV parsing is a direct Python port of
``parseCsvLine``/``parseCsvEntries`` from the TS file so the same
``--key-col``/``--value-col`` semantics apply.
"""

from __future__ import annotations

import json
import os
import urllib.parse
from typing import Any, Dict, List, Optional

from ..auth import require_credential, resolve_base_url
from ..errors import CliError
from ..flags import GlobalOptions, parse_flexible_flag_args
from ..http import call_api
from ._common import parse_json_positional


def _parse_csv_line(line: str) -> List[str]:
    cells: List[str] = []
    current: List[str] = []
    in_quotes = False
    i = 0
    length = len(line)
    while i < length:
        char = line[i]
        nxt = line[i + 1] if i + 1 < length else ""
        if char == '"':
            if in_quotes and nxt == '"':
                current.append('"')
                i += 2
                continue
            in_quotes = not in_quotes
            i += 1
            continue
        if char == "," and not in_quotes:
            cells.append("".join(current))
            current = []
            i += 1
            continue
        current.append(char)
        i += 1
    cells.append("".join(current))
    return [c.strip() for c in cells]


def _parse_csv_entries(
    file_path: str, key_column: str, value_column: Optional[str]
) -> List[Dict[str, Any]]:
    with open(file_path, "r", encoding="utf-8") as fh:
        raw = fh.read()
    lines = [line.rstrip() for line in raw.replace("\r\n", "\n").split("\n") if line.strip()]
    if len(lines) < 2:
        raise CliError("CSV file must contain a header row and at least one data row")

    headers = _parse_csv_line(lines[0])
    try:
        key_index = headers.index(key_column)
    except ValueError as err:
        raise CliError(f"CSV key column not found: {key_column}") from err

    value_index = -1
    if value_column:
        try:
            value_index = headers.index(value_column)
        except ValueError as err:
            raise CliError(f"CSV value column not found: {value_column}") from err

    entries: List[Dict[str, Any]] = []
    for line in lines[1:]:
        cells = _parse_csv_line(line)
        if key_index >= len(cells) or not cells[key_index]:
            raise CliError("CSV row is missing a key value")
        key = cells[key_index]

        if value_column and value_index >= 0:
            value: Dict[str, Any] = {
                "value": cells[value_index] if value_index < len(cells) else "",
            }
        else:
            value = {}
            for column_index, header in enumerate(headers):
                if column_index == key_index:
                    continue
                value[header] = cells[column_index] if column_index < len(cells) else ""

        entries.append({"key": key, "value": value})

    return entries


def _parse_number(value: Optional[str], *, field: str) -> Optional[float]:
    if value is None:
        return None
    try:
        if "." in value or "e" in value or "E" in value:
            return float(value)
        return int(value)
    except ValueError as err:
        raise CliError(f"Option --{field} must be numeric, got: {value!r}") from err


def _strip_none(payload: Dict[str, Any]) -> Dict[str, Any]:
    return {key: value for key, value in payload.items() if value is not None}


def handle_data_cache(
    subcommand: Optional[str],
    rest: List[str],
    *,
    globals_: GlobalOptions,
) -> Any:
    base_url = resolve_base_url(globals_.base_url)
    token = require_credential(globals_.token)

    if subcommand == "list":
        return call_api("GET", "/api/v1/data-caches", base_url=base_url, token=token)

    if subcommand == "create":
        if not rest or rest[0].startswith("--"):
            raise CliError("Missing data-cache slug")
        slug = rest[0]
        parsed = parse_flexible_flag_args(rest[1:], boolean_keys={"json"})
        name = parsed.options.get("name")
        if not name:
            raise CliError("Missing required option: --name")
        key_type = parsed.options.get("key-type")
        if not key_type:
            raise CliError("Missing required option: --key-type")

        value_schema = None
        if parsed.options.get("value-schema"):
            value_schema = parse_json_positional(
                parsed.options["value-schema"], "data-cache value schema"
            )
        fetcher_config = None
        if parsed.options.get("fetcher-config"):
            fetcher_config = parse_json_positional(
                parsed.options["fetcher-config"], "data-cache fetcher config"
            )
        default_ttl_days = _parse_number(parsed.options.get("ttl-days"), field="ttl-days")

        payload = _strip_none(
            {
                "slug": slug,
                "displayName": name,
                "description": parsed.options.get("description"),
                "keyType": key_type,
                "valueSchema": value_schema,
                "fetcherConfig": fetcher_config,
                "defaultTtlDays": default_ttl_days,
            }
        )
        return call_api(
            "POST",
            "/api/v1/data-caches",
            base_url=base_url,
            token=token,
            json_body=payload,
        )

    if subcommand == "get":
        if not rest or rest[0].startswith("--"):
            raise CliError("Missing data-cache slug")
        slug = rest[0]
        return call_api(
            "GET",
            f"/api/v1/data-caches/{urllib.parse.quote(slug, safe='')}",
            base_url=base_url,
            token=token,
        )

    if subcommand == "lookup":
        if len(rest) < 2 or rest[0].startswith("--") or rest[1].startswith("--"):
            raise CliError("Usage: data-cache lookup <slug> <key> --strategy <strategy>")
        slug = rest[0]
        key = rest[1]
        parsed = parse_flexible_flag_args(rest[2:], boolean_keys={"json"})
        strategy = parsed.options.get("strategy")
        if not strategy:
            raise CliError("Missing required option: --strategy")
        payload = _strip_none(
            {
                "key": key,
                "strategy": strategy,
                "maxStalenessWindow": _parse_number(
                    parsed.options.get("max-staleness-days"),
                    field="max-staleness-days",
                ),
                "fetchTimeoutMs": _parse_number(
                    parsed.options.get("fetch-timeout-ms"),
                    field="fetch-timeout-ms",
                ),
            }
        )
        return call_api(
            "POST",
            f"/api/v1/data-caches/{urllib.parse.quote(slug, safe='')}/lookup",
            base_url=base_url,
            token=token,
            json_body=payload,
        )

    if subcommand == "upsert":
        if len(rest) < 2 or rest[0].startswith("--") or rest[1].startswith("--"):
            raise CliError("Usage: data-cache upsert <slug> <key> --value <json>")
        slug = rest[0]
        key = rest[1]
        parsed = parse_flexible_flag_args(rest[2:], boolean_keys={"json"})
        raw_value = parsed.options.get("value")
        if raw_value is None:
            raise CliError("Missing required option: --value")
        value = parse_json_positional(raw_value, "data-cache entry value")
        payload = _strip_none(
            {
                "key": key,
                "value": value,
                "expiresAt": parsed.options.get("expires-at"),
            }
        )
        return call_api(
            "POST",
            f"/api/v1/data-caches/{urllib.parse.quote(slug, safe='')}/entries",
            base_url=base_url,
            token=token,
            json_body=payload,
        )

    if subcommand == "import":
        if not rest or rest[0].startswith("--"):
            raise CliError("Usage: data-cache import <slug> --file <path>")
        slug = rest[0]
        parsed = parse_flexible_flag_args(rest[1:], boolean_keys={"json"})
        file_path = parsed.options.get("file")
        if not file_path:
            raise CliError("Missing --file option")

        ext = os.path.splitext(file_path)[1].lower()
        entries: List[Dict[str, Any]]
        if ext == ".json":
            with open(file_path, "r", encoding="utf-8") as fh:
                raw = fh.read()
            parsed_json = parse_json_positional(raw, "data-cache import file")
            if isinstance(parsed_json, list):
                entries = list(parsed_json)
            elif isinstance(parsed_json, dict) and isinstance(parsed_json.get("entries"), list):
                entries = list(parsed_json["entries"])
            else:
                raise CliError(
                    "JSON import file must be either an array of entries or "
                    "an object with an `entries` array"
                )
        elif ext == ".csv":
            key_col = parsed.options.get("key-col")
            if not key_col:
                raise CliError("CSV imports require --key-col")
            entries = _parse_csv_entries(
                file_path, key_col, parsed.options.get("value-col")
            )
        else:
            raise CliError(
                f"Unsupported import file type: {ext or 'unknown'}"
            )

        return call_api(
            "POST",
            f"/api/v1/data-caches/{urllib.parse.quote(slug, safe='')}/import",
            base_url=base_url,
            token=token,
            json_body={"entries": entries},
        )

    raise CliError(f"Unknown data-cache subcommand: {subcommand or '(none)'}")

File: business-api/src/cli/commands/data-cache.ts
import fs from "node:fs";
import path from "node:path";

import {
  bulkImport as bulkImportDataCacheEntries,
  createCache,
  getCache,
  listCaches,
  lookup as lookupDataCache,
  upsertEntry as upsertDataCacheEntry,
} from "../../services/data-caches.js";
import {
  dataCacheBulkImportSchema,
  dataCacheEntryUpsertSchema,
  dataCacheInputSchema,
  dataCacheLookupSchema,
  type JsonObject,
} from "../../schemas/data-caches.js";
import { parseFlagArgs, parseJsonArg, throwUnknownCommand, type CliCommandDefinition } from "../core.js";

function parseCsvLine(line: string): string[] {
  const cells: string[] = [];
  let current = "";
  let inQuotes = false;

  for (let index = 0; index < line.length; index += 1) {
    const char = line[index];
    const next = line[index + 1];

    if (char === "\"") {
      if (inQuotes && next === "\"") {
        current += "\"";
        index += 1;
      } else {
        inQuotes = !inQuotes;
      }
      continue;
    }

    if (char === "," && !inQuotes) {
      cells.push(current);
      current = "";
      continue;
    }

    current += char;
  }

  cells.push(current);
  return cells.map((cell) => cell.trim());
}

function parseCsvEntries(filePath: string, keyColumn: string, valueColumn?: string): Array<{ key: string; value: JsonObject }> {
  const raw = fs.readFileSync(filePath, "utf8");
  const lines = raw
    .split(/\r?\n/)
    .map((line) => line.trimEnd())
    .filter((line) => line.length > 0);

  if (lines.length < 2) {
    throw new Error("CSV file must contain a header row and at least one data row");
  }

  const headers = parseCsvLine(lines[0]);
  const keyIndex = headers.indexOf(keyColumn);
  if (keyIndex === -1) {
    throw new Error(`CSV key column not found: ${keyColumn}`);
  }

  const valueIndex = valueColumn ? headers.indexOf(valueColumn) : -1;
  if (valueColumn && valueIndex === -1) {
    throw new Error(`CSV value column not found: ${valueColumn}`);
  }

  return lines.slice(1).map((line) => {
    const cells = parseCsvLine(line);
    const key = cells[keyIndex];
    if (!key) {
      throw new Error("CSV row is missing a key value");
    }

    if (valueColumn && valueIndex >= 0) {
      return {
        key,
        value: {
          value: cells[valueIndex] ?? "",
        },
      };
    }

    const value = headers.reduce<Record<string, string>>((accumulator, header, index) => {
      if (index === keyIndex) {
        return accumulator;
      }

      accumulator[header] = cells[index] ?? "";
      return accumulator;
    }, {});

    return { key, value };
  });
}

export const commandDefinitions: CliCommandDefinition[] = [
  {
    scope: "data-cache",
    aliases: ["data-caches"],
    help: {
      description: "Manage persistent reference-data caches and return generic agent instructions for missing values.",
      commands: [
        "list",
        "create <slug> --name <display-name> --key-type <type> [--description <text>] [--value-schema <json>] [--fetcher-config <json>] [--ttl-days <days>]",
        "get <slug>",
        "lookup <slug> <key> --strategy <strategy> [--max-staleness-days <days>] [--fetch-timeout-ms <ms>]",
        "upsert <slug> <key> --value <json> [--expires-at <iso-datetime>]",
        "import <slug> --file <path> [--key-col <name>] [--value-col <name>]",
      ],
      examples: [
        "data-cache list",
        'data-cache create currency-rates-eur-usd --name "Currency Rates EUR/USD" --key-type date --value-schema \'{"type":"object","properties":{"rate":{"type":"string"}},"required":["rate"]}\' --fetcher-config \'{"prompt":"Look up EUR/USD rate for {{ key }}. JSON only."}\' --ttl-days 1',
        "data-cache lookup currency-rates-eur-usd 2026-04-26 --strategy staleness_window --max-staleness-days 7",
        'data-cache upsert currency-rates-eur-usd 2026-04-26 --value \'{"rate":"1.0823"}\'',
      ],
    },
    async handler({ subcommand, rest, positionalArgs, context }) {
      if (subcommand === "list") {
        context.printJson(listCaches());
        return;
      }

      if (subcommand === "create") {
        const slug = rest[0];
        if (!slug) {
          throw new Error("Missing cache slug");
        }

        const { options } = parseFlagArgs(rest.slice(1));
        const input = dataCacheInputSchema.parse({
          slug,
          displayName: options.name,
          description: options.description,
          keyType: options["key-type"],
          valueSchema: options["value-schema"] ? parseJsonArg(options["value-schema"], "data-cache value schema") : undefined,
          fetcherConfig: options["fetcher-config"] ? parseJsonArg(options["fetcher-config"], "data-cache fetcher config") : undefined,
          defaultTtlDays: options["ttl-days"] ? Number(options["ttl-days"]) : undefined,
        });
        context.printJson(createCache(input));
        return;
      }

      if (subcommand === "get") {
        context.printJson(getCache(rest[0]));
        return;
      }

      if (subcommand === "lookup") {
        const slug = rest[0];
        const key = rest[1];
        if (!slug || !key) {
          throw new Error("Usage: data-cache lookup <slug> <key> --strategy <strategy>");
        }

        const { options } = parseFlagArgs(rest.slice(2));
        const input = dataCacheLookupSchema.parse({
          key,
          strategy: options.strategy,
          maxStalenessWindow: options["max-staleness-days"] ? Number(options["max-staleness-days"]) : undefined,
          fetchTimeoutMs: options["fetch-timeout-ms"] ? Number(options["fetch-timeout-ms"]) : undefined,
        });
        context.printJson(await lookupDataCache(slug, input.key, input));
        return;
      }

      if (subcommand === "upsert") {
        const slug = rest[0];
        const key = rest[1];
        if (!slug || !key) {
          throw new Error("Usage: data-cache upsert <slug> <key> --value <json>");
        }

        const { options } = parseFlagArgs(rest.slice(2));
        const input = dataCacheEntryUpsertSchema.parse({
          key,
          value: parseJsonArg(options.value, "data-cache entry value"),
          expiresAt: options["expires-at"],
        });
        context.printJson(upsertDataCacheEntry(slug, input.key, input.value, "manual", input.expiresAt));
        return;
      }

      if (subcommand === "import") {
        const slug = rest[0];
        if (!slug) {
          throw new Error("Usage: data-cache import <slug> --file <path>");
        }

        const { options } = parseFlagArgs(rest.slice(1));
        const filePath = options.file;
        if (!filePath) {
          throw new Error("Missing --file option");
        }

        const ext = path.extname(filePath).toLowerCase();
        let entries: Array<{ key: string; value: JsonObject; expiresAt?: string }>;

        if (ext === ".json") {
          const parsed = parseJsonArg(fs.readFileSync(filePath, "utf8"), "data-cache import file");
          entries = Array.isArray(parsed)
            ? dataCacheBulkImportSchema.parse({ entries: parsed }).entries
            : dataCacheBulkImportSchema.parse(parsed).entries;
        } else if (ext === ".csv") {
          if (!options["key-col"]) {
            throw new Error("CSV imports require --key-col");
          }

          entries = parseCsvEntries(filePath, options["key-col"], options["value-col"]);
        } else {
          throw new Error(`Unsupported import file type: ${ext || "unknown"}`);
        }

        context.printJson(bulkImportDataCacheEntries(slug, entries));
        return;
      }

      throwUnknownCommand(positionalArgs);
    },
  },
];
